### ISSUE TITLE
Reproducibility of LSUN_Churches FID socre

### ISSUE THREAD

**Issue by ader47** [CompVis/latent-diffusion#253]
I sampled 50k images using the ckeckpoint on the LSUN_Churches dataset provided in the repo, calculated the FID score via torch-fidelity, I try both config of sampler :
1. 400 DDIM steps, eta=0. (the repo settting)
2. 200 DDIM steps, eta=1. 

But the FID score is much worse than the results reported in the original paper.
400 DDIM steps FID score is 14.063717185112182
200 DDIM steps FID score is 15.891808501789257   
The FID score in the paper is 4.02

Is anyone had this issue?? And why would this happen?

**Comment by quandao10** [comment]
Hi @ader47, can you reproduce the result ?. I'm still struggling with it

**Comment by ader47** [comment]
> Hi @ader47, can you reproduce the result ?. I'm still struggling with it

yes, I can produce the FID with the checkpoint provided, but I can’t reproduce the FID with my own trained checkpoint. I trained with 2 cards and the FID was much worse than the FID reported.

**Comment by santisy** [comment]
> > Hi @ader47, can you reproduce the result ?. I'm still struggling with it
> 
> yes, I can produce the FID with the checkpoint provided, but I can’t reproduce the FID with my own trained checkpoint. I trained with 2 cards and the FID was much worse than the FID reported.

Hi, may I know what was the original cause that you could not reproduce the FID?

**Comment by ader47** [comment]
> 

I still can not reproduce the FID with my own trained checkpoint.
The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.

**Comment by santisy** [comment]
> > 
> 
> I still can not reproduce the FID with my own trained checkpoint. The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.

Thank you for your reply. Very helpful.

**Comment by eslambakr** [comment]
Hi @ader47 ,
I also training the unconditional model from scratch on the LSUN-churches dataset using the default setup using 4 GPUs.
And after trained for 100 epochs, the sampled images are really terrible.
Here is a sample using the last checkpoint (Epoch 100)
![sample_000186](https://github.com/CompVis/latent-diffusion/assets/25908688/dcb8d40a-7fd9-4fe0-9630-4d86281eb58f)
Here is the result for the best checkpoint (Epoch 53)
![sample_000001](https://github.com/CompVis/latent-diffusion/assets/25908688/8dadbc7d-354f-44c0-be67-b428c975252b)
Accordingly, I didn't calculate the FID as there is no need :D 
The qualitative results are enough :D 

Anyone managed to reproduce the results or even get a good ones using any dataset?

**Comment by eslambakr** [comment]
Hi guys,
Following the issue I just reported.
I wanna share with u this info so it may be useful for others.
Initially when I trained the model on churches datasets I got random and balck images which indicates that the model didn't learn anything.
The solution proposed by [rromb](https://github.com/CompVis/latent-diffusion/issues/8#issuecomment-1141479714) ; deactivate the learning rate scaling `--scale_lr False`, works for me and now I have a good images after 70 epochs and the training is on going.
![sample_000002](https://github.com/CompVis/latent-diffusion/assets/25908688/24969f9e-b25b-4a75-b941-e355dfa1e13d)
![sample_000006](https://github.com/CompVis/latent-diffusion/assets/25908688/e7ee6928-eac9-44a1-9954-cb7b36e1c276)

**Comment by chiendo1010** [comment]
> Hi guys, Following the issue I just reported. I wanna share with u this info so it may be useful for others. Initially when I trained the model on churches datasets I got random and balck images which indicates that the model didn't learn anything. The solution proposed by [rromb](https://github.com/CompVis/latent-diffusion/issues/8#issuecomment-1141479714) ; deactivate the learning rate scaling `--scale_lr False`, works for me and now I have a good images after 70 epochs and the training is on going. ![sample_000002](https://user-images.githubusercontent.com/25908688/244863998-24969f9e-b25b-4a75-b941-e355dfa1e13d.png) ![sample_000006](https://user-images.githubusercontent.com/25908688/244864011-e7ee6928-eac9-44a1-9954-cb7b36e1c276.png)

I would like to ask how to generate church image from [pretrain weight](https://github.com/CompVis/latent-diffusion#pretrained-ldms)

**Comment by Alii-Ganjj** [comment]
> > 
> 
> I still can not reproduce the FID with my own trained checkpoint. The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.

Hi @ader47, I have a few questions in this regard. I highly appreciate it if you help me.

1. Can you please provide more details about how could you reproduce the FID scores? 
2. I have sampled 50k images from my model using the `[sample_diffusion](https://github.com/CompVis/latent-diffusion/blob/main/scripts/sample_diffusion.py)` script provided in the repo.
Now, I don't know what exact code should I execute using the `torch-fidelity` package to calculate FID.
3. The original images that I downloaded and processed from `[lsun](https://github.com/fyu/lsun)` are in `webp` format. Is that the same as yours?

Thanks.

**Comment by ader47** [comment]
> 

I calculate FID via the `torch-fidelity` package, and I downloaded and processed lsun dataset from `https://github.com/fyu/lsun` in `webp` format.

**Comment by eslambakr** [comment]
Hi @ader47,
I still can not reproduce the results even using the provided checkpoint for LSUN dataset.
I am using the sampling code mentioned in the readme 
`CUDA_VISIBLE_DEVICES=<GPU_ID> python scripts/sample_diffusion.py -r models/ldm/<model_spec>/model.ckpt -l <logdir> -n <\#samples> --batch_size <batch_size> -c <\#ddim steps> -e <\#eta> `

I used 50 steps 200 and 400 steps, and here is the results:

1. 50    --> 9 FID
2. 200 --> 7 FID
3. 400 --> 7 FID

Also FYI I am generating only 5k samples not 50k but I think it will not make any difference as visually the generated images are not too good to achieves 4 FID.

So, it would be highly appreciated if u can share with us how could u reproduce the 4 FID, which is reported in the paper.

**Comment by JierunChen** [comment]
> > 
> 
> I still can not reproduce the FID with my own trained checkpoint. The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.

Hi, could you be more specific on pre-processing the images? Do you apply the pre-processing to the original datasets or the sampled images?

**Comment by JierunChen** [comment]
> > > 
> > 
> > 
> > I still can not reproduce the FID with my own trained checkpoint. The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.
> 
> Hi @ader47, could you be more specific on pre-processing the images? Do you apply the pre-processing to the original datasets or the sampled images?

**Comment by ader47** [comment]
> > > > 
> > > 
> > > 
> > > I still can not reproduce the FID with my own trained checkpoint. The provided checkpoint, I didn't pre-process images in the dataset properly. And I can reproduce the FID score after I pre-processed the images with the code in the `LSUN` dataloader.
> > 
> > 
> > Hi @ader47, could you be more specific on pre-processing the images? Do you apply the pre-processing to the original datasets or the sampled images?

yes, I trained the ldm model with the provided yaml config file, it contains the dataset pre-process, and I tested on the pre-processed dataset. I can reproduce the FID score with the provided checkpoint but I can not reproduce with own trained checkpoint.

**Comment by Sugar929** [comment]
@ader47 hello, I want to know when calculate fid:
    1.How did you choose 50k pics from lsun dataset?  how to decide which pic I should choose?
    2.The pictures from dataset is in webp format, how did you use it to calculate fid while generated pictures are in png format?
    Thank u

**Comment by ader47** [comment]
> @ader47 hello, I want to know when calculate fid: 1.How did you choose 50k pics from lsun dataset? how to decide which pic I should choose? 2.The pictures from dataset is in webp format, how did you use it to calculate fid while generated pictures are in png format? Thank u

Hi, don't need to chosse 50k pics from the dataset, unconditonal sampled 50k pics with the trained model, and then calculate the FID with the training set of the dataset. (Paper setting)
I have calculate FID generated pics in both png format and jpg format, and found that png format has lower FID score.
By the way, before calculate FID, the dataset should be preprocessed, follow the training setting in code, resize to 256*256 and save in png format.

**Comment by ader47** [comment]
> @ader47 hello, I want to know when calculate fid: 1.How did you choose 50k pics from lsun dataset? how to decide which pic I should choose? 2.The pictures from dataset is in webp format, how did you use it to calculate fid while generated pictures are in png format? Thank u

I forgot what format I saved the dataset in when I was processing it..... maybe not png format

**Comment by Sugar929** [comment]
@ader47 thanks for your detailed answer, but I still have a question... when we calculate fid score between two directories of pictures, should we set the counts be the same? If that's true, pictures of training set are far more than 50k, it seemed quite weird.

**Comment by ader47** [comment]
> @ader47 thanks for your detailed answer, but I still have a question... when we calculate fid score between two directories of pictures, should we set the counts be the same? If that's true, pictures of training set are far more than 50k, it seemed quite weird.

yes, training dataset is far more than 50k, I think maybe the more data, the more accurate the FID will be, whether it is generated or in the data set.
![image](https://github.com/CompVis/latent-diffusion/assets/46866849/2875bd8d-6fec-4e17-8078-fa6a397b9266)
